src/client_state.js

```diff
@@ -8,11 +8,29 @@ export class ClientState {
 		// show the status for a certain user...
 		// ie online, offline etc
 		this.users = Immutable({});
+		// store which channels contain references to the specified user...
+		this.userChannelReferences = {};
+	}
+
+	updateUsers(users) {
+		for (const user of users) {
+			this.users = this.users.set(user.id, Immutable(user));
+		}
 	}
 
 	updateUser(user) {
 		if (user != null) {
 			this.users = this.users.set(user.id, Immutable(user));
 		}
 	}
+
+	updateUserReference(user, channelID) {
+		this.updateUser(user);
+		if (!this.userChannelReferences[user.id]) {
+			this.userChannelReferences[user.id] = [];
+		}
+		this.userChannelReferences[user.id].push(channelID);
```

Comment:
I might be paranoid here but perhaps we should err on safe side and go for a set instead of array (eg. Object.Keys)

```
if (!this.userChannelReferences[user.id]) {
	this.userChannelReferences[user.id] = {};
}
this.userChannelReferences[user.id][channelID] = true;
```

Reply:
looking at `_initializeState` this might actually be a good idea (updateUserReference is called on each member and watcher)

Reply:
done